src/prototype_room_init.js

```diff
@@ -170,182 +220,140 @@ Room.prototype.setTowerFiller = function() {
-        continue;
-      }
-      if (this.memory.position.structure.lab.length < CONTROLLER_STRUCTURES.lab[8] ||
-        this.memory.position.structure.terminal.length < CONTROLLER_STRUCTURES.terminal[8]) {
-        this.log('Structures not found: ' +
-          'lab: ' + this.memory.position.structure.lab.length + ' ' +
-          'terminal: ' + this.memory.position.structure.terminal.length
-        );
-        continue;
-      }
-      if (!this.memory.position.pathEnd) {
-        this.log('Room not completly build');
-      }
       console.log('All labs/terminal set: ' + pathI);
-      this.setMemoryCostMatrix(costMatrix);
       return pathI;
     }
   }
-  this.setMemoryCostMatrix(costMatrix);
 
   return -1;
 };
 
+Room.prototype.checkPositions = function() {
+  for (let type of ['spawn', 'extension', 'tower', 'link', 'observer', 'nuker']) {
+    if (this.memory.position.structure[type].length < CONTROLLER_STRUCTURES[type][8]) {
+      this.log('Structures not found: ' +
+        'spawns: ' + this.memory.position.structure.spawn.length + ' ' +
+        'extensions: ' + this.memory.position.structure.extension.length + ' ' +
+        'towers: ' + this.memory.position.structure.tower.length + ' ' +
+        'links: ' + this.memory.position.structure.link.length + ' ' +
+        'observer: ' + this.memory.position.structure.observer.length + ' ' +
+        'lab: ' + this.memory.position.structure.lab.length + ' ' +
+        'terminal: ' + this.memory.position.structure.terminal.length + ' ' +
+        'nuker: ' + this.memory.position.structure.nuker.length
+      );
+      return true;
+    }
+  }
+  return false;
+};
+
+Room.prototype.setExtensionPos = function(structurePos, pathI) {
+  this.setPosition('extension', structurePos, config.layout.structureAvoid);
+  if (!this.memory.position.pathEndLevel) {
+    this.memory.position.pathEndLevel = [0];
+  }
+  if (CONTROLLER_STRUCTURES.extension[this.memory.position.pathEndLevel.length] <= this.memory.position.structure.extension.length) {
+    this.memory.position.pathEndLevel.push(pathI);
+  }
+};
+
+Room.prototype.areEnergyStructuresPlaced = function() {
+  return this.memory.position.structure.spawn.length < CONTROLLER_STRUCTURES.spawn[8] && this.memory.position.structure.extension.length < CONTROLLER_STRUCTURES.extension[8];
+};
+
+Room.prototype.setPositionAux = function(structurePos) {
+  for (let type of ['tower', 'nuker', 'observer', 'link']) {
+    if (this.memory.position.structure[type].length < CONTROLLER_STRUCTURES[type][8]) {
+      this.setPosition(type, structurePos, config.layout.structureAvoid);
+      return true;
+    }
+  }
+  return false;
+};
+
+Room.prototype.setStructuresCheck = function(pathI) {
+  if (!this.memory.position.pathEnd) {
+    this.log('Room not completly build');
+  }
+  console.log('All structures set: ' + pathI);
+};
+
+Room.prototype.setStructuresIteratePos = function(structurePos, pathI, path, pathPos) {
```

Comment:
wouldn't it better to replace 2 parameters `path` and `pathI` with one - `pathPosNext`? or jest drop `pathPos` parameter and create it as varable inside function